### ISSUE TITLE
Docs: Add a tutorial for how to use Prefect with a job scheduler on an HPC machine 

### ISSUE THREAD

**Issue by arosen93** [PrefectHQ/prefect#10136]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used GitHub search to find a similar request and didn't find it 😇

### Describe the issue

There are currently no examples for how to use Prefect with a job scheduling system (e.g. SLURM) on an HPC machine. I think this is a pretty important omission because most academics and users of the top supercomputers might not be aware how they can use Prefect.

### Describe the proposed change

Add a tutorial.

### Additional context

There's mention of spinning up a Dask cluster but no representative example of how this is done in practice with a given job scheduling system.

**Comment by alexisthual** [comment]
I'm all in for a tutorial on how to use Prefect with a stack using `submitit`, `hydra`, and possibly `dora`!

**Comment by arosen93** [comment]
I must admit that I have no clue what any of those 3 are 😅

**Comment by Andrew-S-Rosen** [comment]
@JBlaschke: Here is an example I wrote up for Perlmutter.

## Some Utility Functions

```python
def make_dask_runner(
    cluster_kwargs: dict,
    cluster_class: callable = None,
    adapt_kwargs: dict[str, int | None] | None = None,
    client_kwargs: dict = None,
    temporary: bool = False,
) -> DaskTaskRunner:
    """
    Make a DaskTaskRunner for use with Prefect workflows.

    Parameters
    ----------
    cluster_kwargs
        Keyword arguments to pass to `cluster_class`.
    cluster_class
        The Dask cluster class to use. Defaults to `dask_jobqueue.SLURMCluster`.
    adapt_kwargs
        Keyword arguments to pass to `cluster.adapt` of the form `{"minimum": int, "maximum": int}`.
        If `None`, no adaptive scaling will be done.
    client_kwargs
        Keyword arguments to pass to `dask.distributed.Client`.
    temporary
        Whether to use a temporary cluster. If `True`, the cluster will be
        terminated once the `Flow` is finished. If `False`, the cluster will
        run until the walltime is reached and can run multiple `Flow`s.

    Returns
    -------
    DaskTaskRunner
        A DaskTaskRunner object for use with Prefect workflows.
    """
    from dask_jobqueue import SLURMCluster
    from prefect_dask.task_runners import DaskTaskRunner

    if cluster_class is None:
        cluster_class = SLURMCluster

    # Make the one-time-use DaskTaskRunner
    if temporary:
        return DaskTaskRunner(
            cluster_class=cluster_class,
            cluster_kwargs=cluster_kwargs,
            adapt_kwargs=adapt_kwargs,
            client_kwargs=client_kwargs,
        )

    # Make the Dask cluster
    cluster = _make_dask_cluster(cluster_class, cluster_kwargs)

    # Set up adaptive scaling
    if adapt_kwargs and (adapt_kwargs["minimum"] or adapt_kwargs["maximum"]):
        cluster.adapt(minimum=adapt_kwargs["minimum"], maximum=adapt_kwargs["maximum"])

    # Return the DaskTaskRunner with the cluster address
    return DaskTaskRunner(address=cluster.scheduler_address)


def _make_dask_cluster(
    cluster_class: callable, cluster_kwargs: dict, verbose=True
) -> DaskJobqueueJob:
    """
    Make a Dask cluster for use with Prefect workflows.

    Parameters
    ----------
    cluster_class
        The Dask cluster class to use. Defaults to `dask_jobqueue.SLURMCluster`.
    cluster_kwargs
        Keyword arguments to pass to `cluster_class`.
    verbose
        Whether to print the job script to stdout.
    """

    cluster = cluster_class(**cluster_kwargs)
    if verbose:
        print(
            f"Workers are submitted with the following job script:\n{cluster.job_script()}"
        )
        print(f"Scheduler is running at {cluster.scheduler.address}")
        print(f"Dashboard is located at {cluster.dashboard_link}")

    return cluster
```

## Example Usage

```python
n_slurm_jobs = 1 # Number of Slurm jobs to launch in parallel.
n_nodes_per_calc = 1 # Number of nodes to reserve for each Slurm job.
n_cores_per_node = 48 # Number of CPU cores per node.
mem_per_node = "64 GB" # Total memory per node.

cluster_kwargs = {
    # Dask worker options
    "n_workers": n_slurm_jobs,
    "cores": n_cores_per_node,
    "memory": mem_per_node,
    # SLURM options
    "shebang": "#!/bin/bash",
    "account": "AccountName",
    "walltime": "00:10:00",
    "job_mem": "0", # (4)!
    "job_script_prologue": [
        "source ~/.bashrc",
        "conda activate MyEnv",
    ],
    "job_directives_skip": ["-n", "--cpus-per-task"],
    "job_extra_directives": [f"-N {n_nodes_per_calc}", "-q debug", "-C cpu"],
    "python": "python",
}

runner = make_dask_runner(cluster_kwargs, temporary=True)
```

1. Number of Slurm jobs to launch.

2. Total number of cores (per Slurm job) for Dask worker.

3. Total memory (per Slurm job) for Dask worker.

4. Request all memory on the node.

5. Commands to run before calculation. This is a good place to include environment variable definitions and modules to load.

6. Slurm directives that are automatically added but that we chose to skip.

7. The number of nodes for each calculation (-N), queue name (-q), and constraint (-c). Oftentimes, the constraint flag is not needed.

8. The Python executable name. This often does not need to be changed.

With this instantiated cluster object, you can set the task runner of the `Flow` as follows.

```python
@flow(task_runner=runner)
def workflow(atoms):
    ...
```

Now, when the worklow is run from the login node, it will be submitted to the job scheduling system (Slurm by default), and the results will be sent back to Prefect Cloud once completed.

!!! Tip

    Refer to the [Dask-Jobqueue Documentation](https://jobqueue.dask.org/en/latest/generated/dask_jobqueue.SLURMCluster.html) for the available `cluster_kwargs` that can be defined and how they relate to a typical job script.


To asynchronously spawn a Slurm job that continually pulls in work for the duration of its walltime (rather than starting and terminating over the lifetime of the associated `Flow`), you can instead use the `make_dask_runner` command without a `temporary` keyword argument:

```python
runner = make_dask_runner(cluster_kwargs)
```

This is often more efficient for running large numbers of workflows because you can request a single, large Slurm job that continually pulls in work rather than submitting a large number of small jobs to the scheduler.

Additionally, you can have the generated Dask cluster adaptively scale based on the amount of work available by setting `adapt_kwargs` as follows:

```python
runner = make_dask_runner(cluster_kwargs, adapt_kwargs={"minimum": 1, "maximum": 5})
```

This will ensure that at least one Slurm job is always running, but the number of jobs will scale up to 5 if there is enough work available.

**Troubleshooting**

If you are having trouble figuring out the right `cluster_kwargs` to use, the best option is to have the generated job script printed to the screen. This can be done as follows.

```python
from dask_jobqueue import SLURMCluster
from quacc.util.wflows import _make_dask_cluster

cluster = make_cluster(SLURMCluster, cluster_kwargs, verbose=True)
```

Note, however, that a Slurm job will be immediately submitted, so you will probably want to `scancel` it as you debug your job script.

**Executor Configuration File**

Speaking of configurations, if you use mostly the same HPC settings for your calculations, it can be annoying to define a large dictionary in every workflow you run. Instead, you can define a configuration file at `~/.config/dask/jobqueue.yaml` as described in the [dask-jobqueue documentation](https://jobqueue.dask.org/en/latest/configuration-setup.html#managing-configuration-files) that can be used to define default values common to your HPC setup.

**Using a Prefect Work Pool and Agent**

So far, we have dispatched calculations immediately upon calling them. However, in practice, it is often more useful to have a Prefect agent running in the background that will continually poll for work to submit to the task runner. This allows you to submit only a subset of workflows at a time, and the agent will automatically submit more jobs as the resources become available. You will want to run Prefect workflows with an agent on the computing environment where you wish to submit jobs, specifically on a perpetual resource like a login node or dedicated workflow node. Refer to the ["Work Pools, Workers, and Agents"](https://docs.prefect.io/latest/concepts/work-pools/) section of the Prefect documentation for more details.

**Comment by JBlaschke** [comment]
@Andrew-S-Rosen thank you -- I'll dust this off and condense it into a tutorial for our docs.

**Comment by Andrew-S-Rosen** [comment]
Here is a worked example for those curious. It is meant to work on the Perlmutter machine at NERSC.

## Preliminary Steps

On the login node:

```bash
pip install prefect prefect-dask dask-jobqueue
prefect cloud login
```

## Basic Example

Prefect has a Dask and Ray backend, but only Dask (via `dask-jobqueue`) interfaces cleanly with Slurm. We'll use `dask-jobqueue` as the backend for simplicity.

We start by defining the `SLURMCluster()` object in `dask-jobqueue`. Namely, we're outlining what goes in the Slurm submission script.

```python
from dask.distributed import Client
from dask_jobqueue import SLURMCluster

cluster_kwargs = {
    "cores": 128,  
    "memory": "64 GB",
    "shebang": "#!/bin/bash",
    "account": "MyAccountName",
    "walltime": "00:10:00",
    "job_mem": "0",
    "job_script_prologue": ["source ~/.bashrc"],
    "job_directives_skip": ["-n", "--cpus-per-task"],  
    "job_extra_directives": ["-q debug", "-C cpu"],  
}

cluster = SLURMCluster(**cluster_kwargs)
print(cluster.job_script())
```

Now we define how many Slurm jobs we want with those specs and instantie the Dask cluster. This will immediately submit a job to the queue even though we don't have any compute tasks to run just yet.

```python
slurm_jobs = 1
cluster.scale(jobs=slurm_jobs)
client = Client(cluster)
```

Now we'll define our Prefect workflow (`@flow`) and tell it to run the individual `@task`s via our active Dask cluster. This can be done using the `DaskTaskRunner` in Prefect, which allows you to pass an address for the Dask cluster.

```python
from prefect import flow, task
from prefect_dask import DaskTaskRunner

@flow(task_runner=DaskTaskRunner(address=client.scheduler.address))
def workflow(a: float, b: float) -> float:
    output1 = add.submit(a, b)
    output2 = mult.submit(output1, b)
    return output2

@task
def add(a: float, b: float) -> float:
    return a + b

@task
def mult(a: float, b: float) -> float:
    return a * b
```

Now we instantiate and execute the workflow. The progress can be traced in Prefect Cloud.

```python
output = workflow(1, 2)
print(output.result())
```

Since the Dask cluster remains alive until the walltime or it is killed, we can run another workflow if we want.

## Temporary Dask Cluster

Some users may prefect to spin up a Dask cluster (i.e. Slurm job) for each indiviual `@flow`. This is also possible in Prefect. (Note that the example below will complain about an open port if you have already instantiated a Dask cluster above.)

```python
from prefect import flow, task
from prefect_dask import DaskTaskRunner

@flow(task_runner=DaskTaskRunner(cluster_class=SLURMCluster, cluster_kwargs=cluster_kwargs))
def workflow(a: float, b: float) -> float:
    output1 = add.submit(a, b)
    output2 = mult.submit(output1, b)
    return output2

@task
def add(a: float, b: float) -> float:
    return a + b

@task
def mult(a: float, b: float) -> float:
    return a * b

workflow(1, 2).result()
```

At this point, since the `@task` is finished, the Slurm job is no longer running.

## Limitations

### Prefect Cloud is Fine at NERSC But Not Everywhere

Need a network connection from the *compute node*.

### Killed Slurm Jobs Aren't Reflected in the UI

If you `scancel` a Slurm job, it's possible that the `@flow` may be stuck in a running state in the UI.

### Pilot Job Behavior is Limited

The main limitation of Prefect in this kind of setup is that it inherits the limitations of `dask-jobqueue`. Most notably, you can't distribute `@task`s over multiple nodes on a single Slurm allocation. For instance, the following is not possible:
- Running 256 single-core tasks over two 128-core CPU nodes in a single Slurm allocation. `dask-jobqueue` only knows how to do this over 1 node.
- Running two concurrent VASP jobs requiring 1 GPU node each across 2 GPU nodes in a single Slurm allocation. `dask-jobqueue` does not know how to split up the underlying MPI processes over multiple nodes in a single allocation.

If this kind of behavior is critical, you are better off using Parsl or FireWorks.

Some details:
- https://github.com/dask/dask-jobqueue/issues/459
- https://github.com/dask/dask-jobqueue/issues/616

### Concurrent `@flow`s is Not Yet Supported

If you want to run multiple concurrent `@flow`s without blocking in a single Python process, you can't (yet). You have to submit each `@flow` sequentially or in separate Python processes. This is kind of annoying, but there is an open issue about it.

Details:
- https://github.com/PrefectHQ/prefect/issues/6689

**Comment by pgierz** [comment]
Hello,

I'm trying to use this example to get tasks to execute on the actual compute node and running into difficulties. Here is what I have so far:

```python
#!/usr/bin/env python3
"""
Submits work to SLURM via dask_jobqueue.

In this example, we use Dask in conjunction with Prefect to submit a job to SLURM,
and retrieve information about the SLURM node and job specifications. No compute-heavy
work is done here.
"""

import os
import socket

from dask.distributed import Client
from dask_jobqueue import SLURMCluster
from prefect import flow, task
from prefect.logging import get_run_logger
from prefect_dask import DaskTaskRunner

CLUSTER_KWARGS = {
    "account": "computing.computing",
    "walltime": "00:10:00",
    "memory": "8Gb",
    "cores": 1,
    "n_workers": 1,
}


CLUSTER = SLURMCluster(**CLUSTER_KWARGS)
SLURM_JOBS = 1
CLUSTER.scale(jobs=SLURM_JOBS)
CLIENT = Client(CLUSTER)


@task
def show_job_script():
    logger = get_run_logger()
    logger.info(CLUSTER.job_script())


@task
def show_slurm_info():
    logger = get_run_logger()
    logger.info("Finding slurm environment variables...")
    for var_name, var_value in os.environ.items():
        if var_name.startswith("SLURM"):
            logger.info(f"{var_name}={var_value}")


@task
def show_hostname():
    logger = get_run_logger()
    hostname = socket.getfqdn()
    logger.info(f"Running on {hostname}")


@flow(task_runner=DaskTaskRunner(address=CLIENT.scheduler.address))
def display_slurm_info():
    show_job_script()
    show_slurm_info()
    show_hostname()


if __name__ == "__main__":
    display_slurm_info()
```
Here is the corresponding output:
```console
pgierz in 🌐 albedo0 in prefect-examples on  main [?] via 🐍 v3.10.6 (python-3.10.6) took 43s
❯ python prefect-slurm-info.py
09:13:57.780 | INFO    | prefect.engine - Created flow run 'muscular-kingfisher' for flow 'display-slurm-info'
09:13:57.783 | INFO    | prefect.engine - View at http://127.0.0.1:4200/runs/flow-run/fe8cd180-eb51-4c4a-93d5-c815df87e28b
09:13:57.834 | INFO    | prefect.task_runner.dask - Connecting to existing Dask cluster SLURMCluster(d4d0253f, 'tcp://10.100.1.1:38383', workers=0, threads=0, memory=0 B)
09:13:57.877 | INFO    | Task run 'show_job_script-355' - Created task run 'show_job_script-355' for task 'show_job_script'
09:13:57.898 | INFO    | Task run 'show_job_script-355' - #!/usr/bin/env bash

#SBATCH -J dask-worker
#SBATCH -A computing.computing
#SBATCH -n 1
#SBATCH --cpus-per-task=1
#SBATCH --mem=8G
#SBATCH -t 00:10:00

/albedo/home/pgierz/Code/gitlab.awi.de/hpc/tutorials/prefect-examples/.direnv/python-3.10.6/bin/python -m distributed.cli.dask_worker tcp://10.100.1.1:38383 --name dummy-name --nthreads 1 --memory-limit 7.45GiB --nanny --death-timeout 60

09:13:57.901 | INFO    | Task run 'show_job_script-355' - Finished in state Completed()
09:13:57.918 | INFO    | Task run 'show_slurm_info-73d' - Created task run 'show_slurm_info-73d' for task 'show_slurm_info'
09:13:57.922 | INFO    | Task run 'show_slurm_info-73d' - Finding slurm environment variables...
09:13:57.923 | INFO    | Task run 'show_slurm_info-73d' - SLURM_ACCOUNT=computing.computing
09:13:57.927 | INFO    | Task run 'show_slurm_info-73d' - Finished in state Completed()
09:13:57.944 | INFO    | Task run 'show_hostname-cb1' - Created task run 'show_hostname-cb1' for task 'show_hostname'
09:13:57.950 | INFO    | Task run 'show_hostname-cb1' - Running on albedo0
09:13:57.952 | INFO    | Task run 'show_hostname-cb1' - Finished in state Completed()
09:13:57.998 | INFO    | Flow run 'muscular-kingfisher' - Finished in state Completed()
```

From the output, it is evident that I am still getting information from a login node (these are called `albedo0` and `albedo1`), not one of our compute nodes (these are always called `prod-XXX`). From a second terminal running `squeue --me --iterate 1` I can see that a Dask-worker job is started and then stopped, so the connection between Prefect and Dask seems to work fine, but the tasks are still not running in the correct place. 

What am I doing wrong?

**Comment by JBlaschke** [comment]
Hi @pgierz 

Does squeue show your job as running? Did slurm throw an error (most likely in a log somewhere)

Tbh I don't really know how Dask interacts with Slurm behind the scenes -- at NERSC we actually have make sure Dask is configured correctly (eg. https://gitlab.com/NERSC/nersc-notebooks/-/tree/main/perlmutter/dask) in order to interact with the network correctly.

Is there a reason you're using the Dask Slurm cluster? At NERSC we're working on a way to interact with Slurm directly, which might side-step issues like these.

**Comment by pgierz** [comment]
Hi @JBlaschke,

I could see the job running, and the logs don't show anything interesting, just our standard SLURM prolog output.

> At NERSC we're working on a way to interact with Slurm directly, which might side-step issues like these.

I would be **very** interested in something like this, and am also more than happy to contribute, if that is something you would need more manpower for.

**Comment by pgierz** [comment]
If anyone is interested in looking through what we are doing, the examples are here: https://gitlab.awi.de/hpc/tutorials/prefect-examples

**Comment by pgierz** [comment]
Moin @JBlaschke,

any updates on how NERSC is using Prefect directly with SLURM? I am about to start on yet another project that would benefit from such an integration, so I would be curious about any progress.

**Comment by SandyRogers** [comment]
We (a group at EMBL-EBI) are also using a prefect + slurm setup, without dask because in our fair share setup we cannot block the required amount of cluster resources in that way. However, we have not (yet) written an actual slurm task runner. Instead, we have a dedicated slurm submitter node on which flows run, and a specific subflow that we use to dispatch HPC-type work to other slurm nodes (https://github.com/EBI-Metagenomics/emgapi-v2/blob/main/workflows/prefect_utils/slurm_flow.py#L387). This is just using `pyslurm` to submit a job with a (bash-like) job script, and then using `pyslurm` again, periodically, to monitor the job progress/completion. This only works in setups where:
- you can run flows on some kind of submitter node, i.e. run a prefect worker outside of a slurm job
- flows are a mix of very lightweight tasks (run directly on the submitter node), and very heavy non-python tasks (e.g. running a nextflow pipeline, in our case).

We plan to replace the `pyslurm` dispatch with the slurm REST API. Longer term we might package this logic into a slurm operator, if possible.

**Comment by Andrew-S-Rosen** [comment]
Thanks for sharing this, @SandyRogers! My group has its own login node on the HPC clusters, so it sounds like your kind of setup would be compatible with ours. I'll take a look. The `dask-jobqueue` approach, even if you could support it, is really quite limited and not very practical.

**Comment by Samreay** [comment]
Glad to stumble onto this thread. Recently have joined the Supernova Cosmology Project at LBNL and I'm planning on doing what a lot here are and connecting Prefect to NERSC/generic slurm system. I'd be happy to collaborate with anyone to roll something out, but my approach is going to try and mirror the Prefect3 `ECSWorker` as much as possible, in that it would be good to:

1. Have a templated worker which configures the base slurm template
2. Run everything needed as a container

When I've done this using AWS ECS tasks, we've come across the heterogeneous resource problem too, and for us the solution was to break out flows out into homogenous subflows (all existing within the same image), and orchestrate those sections using Prefect's `run_deployment` command (which we optionally wrap as a task itself if needed). In this way we could have a pipeline that looks something like the following:

```python
from prefect import flow, task

@flow
def process_spectra(...):
   ...your tasks here

@flow
def gpu_training_phase(...):
   ...your tasks here...

@flow
def the_full_pipeline(...):
   ... a mix of low resource tasks here with compute heavy sections processed via run_deployment
```

If this is of interest to anyone else, I can make sure when I turn this into a repo to share back or add collaborators.

**Comment by SandyRogers** [comment]
Interesting, @Samreay , thanks. I also initially went down the route of dispatching the slurm job work always via a `run_deployment`. In the end we don't always do that – usually we just run the slurm-job wrapper flow as a connected subflow. But in a couple of cases it is useful, like when we want to trigger work in slurm jobs from a non-prefect part of our system. 

What is your intended container setup? Will you run all prefect-slurm flows in some base singularity image? I guess singularity is more common in HPC setups than docker/podman.

**Comment by Samreay** [comment]
Right now our image-based tasks default to the primary flow repo which contains all our flows and code (and thus most of our flows), but we have separate projects (such as ML model retraining) which have their own images due to heavy dependencies (tensorflow/pytorch), but each registry definition allows the image ECR to be overwritten. I've simplified the code snippet above, but our actual registry takes in a custom pydantic `FlowDeployment` like so:

![Image](https://github.com/user-attachments/assets/76ad1729-ddbf-440d-87fe-df8d5a471484)

And yeah, its a shame I can't just support docker online, but at least things like Apptainer can run docker images these days, so I'm going to see how far the docker image gets and add the exection service (docker/apptainer/etc) to the worker template. The idea would then be to have the worker able to have any specific variable default overwritten, but also provide a selection of existing configurations set in the cli via `./slurm_worker --config=nersc` and picked up by `pydantic-settings` to make this a simple process and encourage contribution.

The reason I'm leaning towards a worker here is so orchestration can be done via webhook. Ie when someone releases a new version of a stage in the data reduction pipeline, this can trigger a prefect Event, and a long lived `slurm_worker` should be able to automatically kick off whatever orchestration is needed to reprocess the relevant upstream files/processes.

**Comment by colligant** [comment]
> Hello,
> 
> I'm trying to use this example to get tasks to execute on the actual compute node and running into difficulties. Here is what I have so far:
> 
> #!/usr/bin/env python3
> """
> Submits work to SLURM via dask_jobqueue.
> 
> In this example, we use Dask in conjunction with Prefect to submit a job to SLURM,
> and retrieve information about the SLURM node and job specifications. No compute-heavy
> work is done here.
> """
> 
> import os
> import socket
> 
> from dask.distributed import Client
> from dask_jobqueue import SLURMCluster
> from prefect import flow, task
> from prefect.logging import get_run_logger
> from prefect_dask import DaskTaskRunner
> 
> CLUSTER_KWARGS = {
>     "account": "computing.computing",
>     "walltime": "00:10:00",
>     "memory": "8Gb",
>     "cores": 1,
>     "n_workers": 1,
> }
> 
> 
> CLUSTER = SLURMCluster(**CLUSTER_KWARGS)
> SLURM_JOBS = 1
> CLUSTER.scale(jobs=SLURM_JOBS)
> CLIENT = Client(CLUSTER)
> 
> 
> @task
> def show_job_script():
>     logger = get_run_logger()
>     logger.info(CLUSTER.job_script())
> 
> 
> @task
> def show_slurm_info():
>     logger = get_run_logger()
>     logger.info("Finding slurm environment variables...")
>     for var_name, var_value in os.environ.items():
>         if var_name.startswith("SLURM"):
>             logger.info(f"{var_name}={var_value}")
> 
> 
> @task
> def show_hostname():
>     logger = get_run_logger()
>     hostname = socket.getfqdn()
>     logger.info(f"Running on {hostname}")
> 
> 
> @flow(task_runner=DaskTaskRunner(address=CLIENT.scheduler.address))
> def display_slurm_info():
>     show_job_script()
>     show_slurm_info()
>     show_hostname()
> 
> 
> if __name__ == "__main__":
>     display_slurm_info()
> Here is the corresponding output:
> 
> pgierz in 🌐 albedo0 in prefect-examples on  main [?] via 🐍 v3.10.6 (python-3.10.6) took 43s
> ❯ python prefect-slurm-info.py
> 09:13:57.780 | INFO    | prefect.engine - Created flow run 'muscular-kingfisher' for flow 'display-slurm-info'
> 09:13:57.783 | INFO    | prefect.engine - View at http://127.0.0.1:4200/runs/flow-run/fe8cd180-eb51-4c4a-93d5-c815df87e28b
> 09:13:57.834 | INFO    | prefect.task_runner.dask - Connecting to existing Dask cluster SLURMCluster(d4d0253f, 'tcp://10.100.1.1:38383', workers=0, threads=0, memory=0 B)
> 09:13:57.877 | INFO    | Task run 'show_job_script-355' - Created task run 'show_job_script-355' for task 'show_job_script'
> 09:13:57.898 | INFO    | Task run 'show_job_script-355' - #!/usr/bin/env bash
> 
> #SBATCH -J dask-worker
> #SBATCH -A computing.computing
> #SBATCH -n 1
> #SBATCH --cpus-per-task=1
> #SBATCH --mem=8G
> #SBATCH -t 00:10:00
> 
> /albedo/home/pgierz/Code/gitlab.awi.de/hpc/tutorials/prefect-examples/.direnv/python-3.10.6/bin/python -m distributed.cli.dask_worker tcp://10.100.1.1:38383 --name dummy-name --nthreads 1 --memory-limit 7.45GiB --nanny --death-timeout 60
> 
> 09:13:57.901 | INFO    | Task run 'show_job_script-355' - Finished in state Completed()
> 09:13:57.918 | INFO    | Task run 'show_slurm_info-73d' - Created task run 'show_slurm_info-73d' for task 'show_slurm_info'
> 09:13:57.922 | INFO    | Task run 'show_slurm_info-73d' - Finding slurm environment variables...
> 09:13:57.923 | INFO    | Task run 'show_slurm_info-73d' - SLURM_ACCOUNT=computing.computing
> 09:13:57.927 | INFO    | Task run 'show_slurm_info-73d' - Finished in state Completed()
> 09:13:57.944 | INFO    | Task run 'show_hostname-cb1' - Created task run 'show_hostname-cb1' for task 'show_hostname'
> 09:13:57.950 | INFO    | Task run 'show_hostname-cb1' - Running on albedo0
> 09:13:57.952 | INFO    | Task run 'show_hostname-cb1' - Finished in state Completed()
> 09:13:57.998 | INFO    | Flow run 'muscular-kingfisher' - Finished in state Completed()
> From the output, it is evident that I am still getting information from a login node (these are called `albedo0` and `albedo1`), not one of our compute nodes (these are always called `prod-XXX`). From a second terminal running `squeue --me --iterate 1` I can see that a Dask-worker job is started and then stopped, so the connection between Prefect and Dask seems to work fine, but the tasks are still not running in the correct place.
> 
> What am I doing wrong?

@pgierz Did you ever resolve this? I'm running into a similar issue.

**Comment by rjdenham** [comment]
@pgierz and @colligan I'm also experiencing this  on a cluster using PBSPro scheduler. Would be keen to hear any updates.

**Comment by colligant** [comment]
@rjdenham @pgierz this code runs the matmul/sums on a compute node: 
```
cluster = SLURMCluster(**cluster_kwargs)
cluster.scale(1)
cluster.wait_for_workers(1)
client = Client(cluster)

@task
def add(a, b):
    return a + b

@task
def mult(a, b):
    return a @ b

@task
def show_hostname():
    logger = get_run_logger()
    hostname = socket.getfqdn()
    logger.info(f"Running on {hostname}")

@flow(name="matmul", task_runner=DaskTaskRunner(address=client.scheduler.address))
def workflow(mats_a, mats_b):
    show_hostname()
    sums  = add.map(mats_a, mats_b)           # 10 parallel adds
    prods = mult.map(sums, mats_b)            # 10 parallel multiplies
    show_hostname()
    return [prod.result().sum() for prod in prods]

if __name__ == "__main__":
    N = 100                                    # fan-out width
    A = [np.random.rand(100, 100) for _ in range(N)]
    B = [np.random.rand(100, 100) for _ in range(N)]
    results = workflow(A, B)
    print(results)
    time.sleep(10)
```
However, the show_hostname() task emits the hostname of the head node (not a compute node).
A simpler example (below) 
```
cluster = SLURMCluster(**cluster_kwargs)
cluster.scale(1)
cluster.wait_for_workers(1)
client = Client(cluster)

def show_hostname(i):
    hostname = socket.getfqdn()
    print(f"Running on {hostname}, task {i}")
    
client.map([show_hostname(i) for i in range(20)])
```
still emits the name of the head node, implying that the issue is with dask's SlurmCluster/Client 

A cursory google search hasn't revealed the solution. For now I say we look for a better place to raise this issue.

**Comment by colligant** [comment]
@rjdenham @pgierz Running this (omitting the setup code) correctly emits the name of the compute node:

```
cluster = SLURMCluster(**cluster_kwargs)
cluster.scale(1)
cluster.wait_for_workers(1)
client = Client(cluster)

@task
def show_hostname(placeholder):
    hostname = socket.getfqdn()
    return hostname

@flow(name="test", task_runner=DaskTaskRunner(address=client.scheduler.address))
def workflow(tests):
    prods = show_hostname.map(tests)
    return [prod.result() for prod in prods]

if __name__ == "__main__":
    a = range(10)
    results = workflow(a)
    print(results)
    time.sleep(10)
```

So there's something internal to get_run_logger() that's not working as expected, though I'm hesitant to say it's a bug. [Docs](https://docs-2.prefect.io/latest/api-ref/prefect/logging/loggers/#prefect.logging.loggers.get_run_logger). 
I tried the same pattern with a CPU bound task and it ran fine on the compute node.

**Comment by Samreay** [comment]
In case it becomes useful for anyone, I have a slurm worker that's running on NERSC right now and using slurm + podman for orchestrating jobs from Prefect Cloud. Its still early days (it's working but I haven't done observability, graceful crashing, job monitoring, etc), but the repo is based off the docker worker and available here: https://github.com/SNIFS-science/prefect-hpc-worker

**Comment by SandyRogers** [comment]
Hello all, you may also be interested in a package @arajkovic and I have in development and early use at EMBL-EBI to make a prefect worker that dispatches to slurm via the slurm REST API: https://pypi.org/project/prefect-slurm/

**Comment by Andrew-S-Rosen** [comment]
@SandyRogers Do you have any plans to make https://github.com/EBI-metagenomics/prefect-slurm/ public? Also, what are the terms of the license? Without this information, it is admittedly difficult to get value from the package in its current state despite its promise.

**Comment by SandyRogers** [comment]
> @SandyRogers Do you have any plans to make https://github.com/EBI-metagenomics/prefect-slurm/ public? Also, what are the terms of the license? Without this information, it is admittedly difficult to get value from the package in its current state despite its promise.

Yes sorry - we will make that public! It is Apache 2 licensed.

**Comment by Andrew-S-Rosen** [comment]
@SandyRogers --- Any ideas for a timeline? We are looking to do something similar here, and we would much rather contribute to an existing repo than start from scratch if possible.

**Comment by SandyRogers** [comment]
@Andrew-S-Rosen  https://github.com/EBI-Metagenomics/prefect-slurm the repo is now public. I hope this proves useful – contributions very welcome! We are still testing this by the way, but are aiming to be using it in production very soon.

**Comment by desertaxle** [comment]
@SandyRogers Would you be interested in `prefect-slurm` being included on the [Prefect integrations](https://docs.prefect.io/integrations/integrations) page? I can handle updating the page, but I wanted to check to make sure you're cool with that first!

**Comment by SandyRogers** [comment]
@desertaxle that would be great, thank you. We are using this in an internal production setup now so it can expect some maintenance from us at least - keeping up to date with new Slurm and Prefect releases once we move to them. Community input eg additional features bug fixes etc are welcome too.

**Comment by andrewcolinhunt** [comment]
@SandyRogers @desertaxle 

I built another package in this space: [prefect-submitit](https://github.com/dexterity-systems/prefect-submitit). It works differently from EBI's [prefect-slurm](https://github.com/EBI-Metagenomics/prefect-slurm), so here's a quick comparison.

prefect-slurm (EBI) is a Prefect worker. Each flow run becomes one SLURM job, submitted through the SLURM REST API using the standard deployments and work pools setup.

prefect-submitit is a task runner. The flow runs on any machine that can submit to SLURM, and each task becomes its own SLURM job, with task.map() turning into a SLURM job array. It uses [submitit](https://github.com/facebookincubator/submitit) under the hood, so it just calls sbatch. No REST API or deployments required. Individual tasks can request their own resources (one task gets a GPU, another gets 4 CPUs), tasks can run as steps inside an existing allocation, and there's a local mode for development without a cluster.

Short version: Use EBI's worker if you want whole flows submitted to SLURM through work pools. Use prefect-submitit if you want individual tasks within a flow scheduled as SLURM jobs. The two aren't really competing; they just operate at different levels.

I wrote this for my work while at the University of Washington, and it worked on our Slurm cluster there. I have since moved on from that role, but I believe @Andrew-S-Rosen and his group are actively using it on their cluster.
